xls/integrator/ir_integrator.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+#include "xls/integrator/ir_integrator.h"
+
+#include "xls/ir/ir_parser.h"
+
+namespace xls {
+
+// Return the name of 'function', prepended with the 'package' name.
+std::string GetParsableQualifiedFunctionName(Package* package,
+                                             const Function* function) {
+  return "PKGzzz" + package->name() + "zzzFNzzz" + function->name();
+}
+
+absl::Status IntegrationBuilder::CopySourcesToIntegrationPackage() {
```

Comment:
Any reason why you need to copy the functions to a new package rather than create new functions in the existing package? And if you do need it, would an xls::Package::Clone method do what you need?

Reply:
Currently, there is a 2-step process:

1.  Copy input packages into duplicate packages, then rename the functions to avoid later name-space collisions.  This is partly a nicety to avoid modifying the input packages.  It also helps to avoid issues that could come up in modifying things in-place in the original package (e.g. orignal has funcA and funcB, where funcA calls funcB.  In the same package, we make copies copy_funcA and copy_funcB.  Unless we traverse and fix the copied functions, copy_funcA calls funcB instead of copy_funcB.  Fixup is more work and will likely require adjustments as other things ie procs, other invocation types etc are added).  This part could be accomplished with a Package::Clone method if it existed.

2.  We copy the renamed functions to a common package.  Here, I am not assuming that the input functions are from the same source package.  The assumption is that functions that will be integrated were developed independently, are not in the same source package, may have name collisions, and were likely scanned in from separate .ir files.  We need to add them to a common package to ensure type compatability.  This part cannot be avoid with a Package::Clone method.

Reply:
I'd guess you'll need a Function::Clone method which takes a map<Function*,Function*> to remap call sites.

Reply:
If we wanted to add new funcitons in the original package, yes. but I think it's better to copy the original package than to add stuff to it.

Reply:
And it looks like we do have a Function::Clone. It just needs to be modified to take a target package and callsite remapping.

Reply:
Done.